perfplot/main.py

```diff
@@ -49,16 +68,26 @@ def __init__(
             self.timings = self.timings[order]
             self.labels = [self.labels[i] for i in order]
             self.colors = [self.colors[i] for i in order]
-        return
+
+        # Set time unit if specified. Allowed values: ("s", "ms", "us", "ns")
+        if time_unit is None:
+            self.time_unit = self._auto_time_unit()
+        elif time_unit in si_time:
+            self.time_unit = time_unit
+        else:
+            raise ValueError("Provided `time_unit` is not valid")
+
+        # Scale timings based on time_unit
+        self._scale_timings()
```

Comment:
Let's not manipulate the timings array. It should _always_ be in the native unit, `ns`. Let's remove `_current_unit`. The scaling only becomes relevant when plotting, so there we could have a function `def scale_from_ns(unit)` or so that _returns_ a new array without modifying `time_data`.

Reply:
Alright agreed. A method ```scale_from_ns()``` now returns a new array without modifying the ```timings``` attribute. To ensure that the timings are not needessly converted, we can use a conditional as implemented before. However, which version would you prefer? I'm learning toward option 1 to not violate [PEP-20]: Explicit is better than Implicit but I'm not sure how clean you'd like the plot function to look. Let me know!

1. ```plot()``` method explicitly performs a conditional check if scaling is required
``` python
    def plot(self):
        _timings = self.scale_from_ns() if self.time_unit != 'ns' else self.timings
        for t, label, color in zip(_timings, self.labels, self.colors):
            self.plotfun(self.n_range, t, label=label, color=color)
        if self.xlabel:
            plt.xlabel(self.xlabel)
        if self.title:
            plt.title(self.title)
        plt.ylabel(f"Runtime [{self.time_unit}]")
        plt.grid(True)
        plt.legend()
        return
``` 

2. The conditional check is performed implicitly by ```scale_from_ns()``` method

``` python 
    def plot(self):
        for t, label, color in zip(self.scale_from_ns(), self.labels, self.colors):
            self.plotfun(self.n_range, t, label=label, color=color)
        if self.xlabel:
            plt.xlabel(self.xlabel)
        if self.title:
            plt.title(self.title)
        plt.ylabel(f"Runtime [{self.time_unit}]")
        plt.grid(True)
        plt.legend()
        return
```

[PEP-20]: https://www.python.org/dev/peps/pep-0020/

Reply:
Option 2 is fine I'd say, but make it a function outside of the class that takes the timings as argument. `scale_from_ns(self.timings)` is a bit clearer imho.

Reply:
If it's outside of the class it will also require the desired output unit such as ```scale_from_ns(self.timings, self.time_unit)```. Instead, if it's still a method it can have the syntax you desire if the timings are an explicit argument.
``` python
    def scale_from_ns(self, timings):
        """ Converts the ``timings`` from nanoseconds into the desired
        :py:attr:`time_unit` after checking if conversion is necessary.

        :rtype: numpy.ndarray
        """
        # Conversion Factor = Current / Desired Magnitude
        factor = si_time["ns"] / si_time[self.time_unit]

        # Scaling timings after converting to float array
        return timings.astype(numpy.float64) * factor
```
Is this fine? Otherwise I'll make it outside of the class :smiley:

Reply:
This `scale_from_ns(ns_data, target_time_unit)` is good. Best put it outside. It's fine without the conditional, too. Actually, the function would simply be
```
 def scale_from_ns(ns_data, target_time_unit):
    return timings * (si_time["ns"] / si_time[target_time_unit])
```
at which point I doubt we actually need a function for it. How about simply inlining it?

Reply:
Ah nice suggestion, I completely forgot that ```timings.astype(numpy.float64)``` is unnecessary now. However, if we inline it, it becomes hard to test the conversion and we'd have to repeat the in-lined conversion in the test, meaning that if the conversion code changes the test might still pass:

``` python
def plot(self):
    scaled_timings = self.timings * (si_time["ns"] / si_time[self.time_unit])
    for t, label, color in zip(scaled_timings, self.labels, self.colors):
        self.plotfun(self.n_range, t, label=label, color=color)
    if self.xlabel:
        plt.xlabel(self.xlabel)
    if self.title:
        plt.title(self.title)
    plt.ylabel(f"Runtime [{self.time_unit}]")
    plt.grid(True)
    plt.legend()
    return
```
``` python
 # Is the timing being converted correctly?
 scaled_timings = data.timings * (si_time["ns"] / si_time[data.time_unit])
 assert scaled_timings == exp_timing
```
Therefore, I suggest keeping it as a function (even as short as it is)

Reply:
>  # Is the timing being converted correctly?

We don't need to test this. It's so simple and obvious, I don't think this will lead to much confusion. No need to be afraid of repetition either, it's just so basic.